File: src/services/RunEvaluator.cs
using StarSmugglerGo.Domain;

namespace StarSmugglerGo.Services;

public sealed class RunEvaluator
{
    public bool IsGameOver(RunState run, DataSnapshot data, EconomyService economyService, TravelService travelService, UpgradeService upgradeService)
    {
        if (!data.PortsById.TryGetValue(run.Player.CurrentPortId, out PortDefinition? currentPort))
        {
            return false;
        }

        int cheapestTravelCost = upgradeService.GetCheapestTravelCostFromPort(run, data, travelService, currentPort);

        if (run.Player.Credits >= cheapestTravelCost)
        {
            return false;
        }

        int sellableValue = economyService.GetSellableCargoValueAtCurrentPort(run, data);
        return run.Player.Credits + sellableValue < cheapestTravelCost;
    }
}

File: internal/services/run_evaluator.go
package services

import "github.com/oberones/star-smuggler-go/internal/domain"

type RunEvaluator struct{}

func (e RunEvaluator) IsGameOver(run domain.RunState, data domain.DataSnapshot, economy EconomyService, travel TravelService) bool {
	currentPort, ok := data.PortsByID[run.Player.CurrentPortID]
	if !ok {
		return false
	}

	cheapestTravelCost := travel.GetCheapestTravelCostFromPort(currentPort, data.Ports)
	if run.Player.Credits >= cheapestTravelCost {
		return false
	}

	sellableValue := economy.GetSellableCargoValueAtCurrentPort(run, data)
	return run.Player.Credits+sellableValue < cheapestTravelCost
}
